code/ship/ship.cpp

```diff
@@ -11921,7 +11922,16 @@ int ship_fire_secondary( object *obj, int allow_swarm )
 
 		int start_slot, end_slot;
 
-		if ( shipp->flags[Ship_Flags::Secondary_dual_fire] && num_slots > 1) {
+		// Checking for the secondary doublefire disabling here in case it slipped, such as being set to double fire in
+		// a previous session If doublefire is set and((player double fire is disabled and this is the player) or 
+		/// (this weapon's doublefire is disabled))
+		if (shipp->flags[Ship_Flags::Secondary_dual_fire] &&
+			((The_mission.ai_profile->flags[AI::Profile_Flags::Disable_player_secondary_doublefire] &&  obj == Player_obj) ||
+			 Weapon_info[swp->secondary_bank_weapons[swp->current_secondary_bank]].wi_flags[Weapon::Info_Flags::No_doublefire])) {
+			shipp->flags.remove(Ship_Flags::Secondary_dual_fire);
+		}
+
+		if (shipp->flags[Ship_Flags::Secondary_dual_fire] && num_slots > 1) {
```

Comment:
It's less about not trusting my code elsewhere as it is not being certain of all the other code in the game.

I have now tested the one corner case I had in mind. If you go from a version of the game/a mod without the flag to one with, you can keep doublefire set from before. It's a very niche case yes, but it is a possible one.  

Also possible in the case of the weapon flag version is having doublefire set on bank 0 and then swapping to a no-doublefire weapon in bank 1. You could have the check in keycontrol when switching secondary banks, but by doing it here you allow the player to cycle past a no-doublefire bank to another bank without losing the doublefire setting. That does come at the cost of not letting the player see immediately that they've switched to a no-doublefire, but without adding some new state it's a pick your poison situation and this is the option I personally think is the better player experience.

Reply:
Again, if you're not certain of the other code in the game then use an Assertion.  An Assertion is meant to guarantee that the code contract behaves as it's expected to.

Why would you ever go from two versions of the mod in the *same* mission?  Ship flags are supposed to be transient, only existing during the lifetime of one mission.  In order to switch mods you need to exit not only the mission but also the game.

I think it's a better player experience to remove the doublefire flag as soon as the player switches to a no-doublefire bank.  Hiding important feedback from the player is a bad thing.

Reply:
I only have the free mental bandwidth to address this point right now, I'll have to circle back for the rest later.

Maybe ship flags are supposed to be transient, but this one is a bit sticky. Based on my testing, there are many ways that a user could start the mission with an invalid doublefire state.

- While developing a mod or mission, after first disallowing doublefire
- When updating to a new version of a mod or mission that didn't disallow doublefire but now does
- When doublefire was allowed in the last played mission but isn't in the next.
- When replaying the same mission but switching from a doublefire allowed missile loadout to a doublefire disallowed loadout.

I haven't rooted through the code to figure out why this state persists between missions yet but my experimental testing indicates it does, at least in tech room missions. I would expect campaign is the same but haven't tested it.

Is an assertion still the right way to handle all these cases? Or perhaps is this persistence something that needs to be hunted down and destroyed in the first place? I'm not sure that change would be welcome from players who are used to the game remembering they want to doublefire rather than having to set it every mission. Maybe there needs to be a doublefire flag validity check during mission start?

Reply:
Ship flags are cleared between missions, and also when you restart missions.  I don't see how a doublefire flag could carry over under any of those circumstances.  If it does, that's a bug.  In that case, upload a simple test mission and I'll take a look.

An assertion is the correct way to handle those cases.  An assertion is expected to never trigger, but it still performs the check in debug builds just in case.

Reply:
Ship flags are cleared yes, but further investigation revealed what I really should have expected in the first place; several things are stored in the player data and restored from it on mission start, the doublefire flag among them. Nothing I can see suggest it's supposed to be tied to a specific mission, just generally remembering the 'cockpit state' as it were of the player's ship as it was at the end of last session.

So, I probably ought to add the flag validity check in or after the function that does that, and then move things around in key control, and finally replace the stuff here for a sanity assert as you suggest. probably a monday project.